near-sdk-macros/src/core_impl/info_extractor/visitor/view_visitor.rs

```diff
@@ -0,0 +1,75 @@
+use super::{BindgenVisitor, InitAttr, ReturnKind, SerializerAttr};
+use crate::core_impl::{MethodKind, Returns, SerializerType, ViewMethod};
+use syn::{spanned::Spanned, Attribute, Error, Receiver, ReturnType};
+
+pub struct ViewVisitor {
+    is_private: bool,
+    result_serializer: SerializerType,
+    returns: Option<Returns>,
+    receiver: Option<Receiver>,
+}
+
+impl Default for ViewVisitor {
+    fn default() -> Self {
+        Self {
+            is_private: Default::default(),
+            result_serializer: SerializerType::JSON,
+            returns: Default::default(),
+            receiver: Default::default(),
+        }
+    }
+}
+
+impl BindgenVisitor for ViewVisitor {
+    fn visit_init_attr(&mut self, attr: &Attribute, _init_attr: &InitAttr) -> syn::Result<()> {
+        Err(Error::new(attr.span(), "View function can't be an init function at the same time."))
+    }
+
+    fn visit_payable_attr(&mut self, attr: &Attribute) -> syn::Result<()> {
+        Err(Error::new(attr.span(), "View function can't be payable."))
+    }
+
+    fn visit_private_attr(&mut self, _attr: &Attribute) -> syn::Result<()> {
+        self.is_private = true;
+        Ok(())
+    }
+
+    fn visit_result_serializer_attr(
+        &mut self,
+        _attr: &Attribute,
+        result_serializer_attr: &SerializerAttr,
+    ) -> syn::Result<()> {
+        self.result_serializer = result_serializer_attr.serializer_type.clone();
+        Ok(())
+    }
+
+    fn visit_receiver(&mut self, receiver: &Receiver) -> syn::Result<()> {
+        self.receiver = Some(receiver.clone());
+        Ok(())
+    }
+
+    fn visit_result(&mut self, handle_result: bool, return_type: &ReturnType) -> syn::Result<()> {
+        self.returns = match return_type {
+            ReturnType::Default => {
+                Some(Returns { original: return_type.clone(), kind: ReturnKind::Default })
+            }
+            ReturnType::Type(_, typ) => Some(Returns {
+                original: return_type.clone(),
+                kind: super::parse_return_kind(typ, handle_result)?,
+            }),
+        };
+        Ok(())
+    }
+
+    fn build(&self) -> syn::Result<MethodKind> {
```

Comment:
nit: why not consume the builder via `build(self)` instead just incase you reuse the builder? That way you don't have to call clone and just claim the builder's items.

Reply:
Yeah I mentioned this in the original post:

> I am not super happy with how the polymorphic aspect of this approach turned out to be (Box<dyn BindgenVisitor> is unmovable, so build() has to clone visitor field values to the end-structure), so feel free to suggest an alternative.

Basically, there is no way to unwrap `Box` because I can't prove that there is only ever a single instance of it.

Reply:
ahh I see, the trait object. You could also do `build(self: Box<Self>)` for the type signature for now, since this is just an internal trait